File: mobile-sdk/packages/android/bodymeasurementsdk/src/main/java/com/bodymeasurement/sdk/BodyMeasurementClient.kt
package com.bodymeasurement.sdk

import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.asRequestBody
import com.google.gson.Gson
import java.io.File
import java.io.IOException

/**
 * Main client for Body Measurement API
 */
class BodyMeasurementClient(
    private val apiKey: String,
    private val baseURL: String = "http://localhost:8000",
    private val debug: Boolean = false
) {
    private val client = OkHttpClient()
    private val gson = Gson()

    /**
     * Process single-person body measurement
     *
     * Example:
     * ```
     * val client = BodyMeasurementClient(apiKey = "YOUR_API_KEY")
     * client.processMeasurement(
     *     imageFile = File("/path/to/image.jpg"),
     *     fitPreference = FitPreference.REGULAR
     * ) { result ->
     *     result.onSuccess { response ->
     *         println("Size: ${response.recommendedSize}")
     *     }.onFailure { error ->
     *         println("Error: ${error.message}")
     *     }
     * }
     * ```
     */
    fun processMeasurement(
        imageFile: File,
        productId: String? = null,
        fitPreference: FitPreference = FitPreference.REGULAR,
        callback: (Result<MeasurementResponse>) -> Unit
    ) {
        log("Processing measurement...")

        val requestBody = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart(
                "file",
                imageFile.name,
                imageFile.asRequestBody("image/jpeg".toMediaType())
            )
            .addFormDataPart("api_key", apiKey)
            .addFormDataPart("fit_preference", fitPreference.value)
            .apply {
                productId?.let { addFormDataPart("product_id", it) }
            }
            .build()

        val request = Request.Builder()
            .url("$baseURL/api/v1/measurements/process")
            .addHeader("X-API-Key", apiKey)
            .post(requestBody)
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onResponse(call: Call, response: Response) {
                response.use {
                    if (!response.isSuccessful) {
                        val errorMsg = parseErrorMessage(response.body?.string())
                        callback(Result.failure(SDKException("API error (${response.code}): $errorMsg")))
                        return
                    }

                    try {
                        val body = response.body?.string()
                        val result = gson.fromJson(body, MeasurementResponse::class.java)
                        log("Measurement successful", result)
                        callback(Result.success(result))
                    } catch (e: Exception) {
                        callback(Result.failure(SDKException("Failed to parse response: ${e.message}")))
                    }
                }
            }

            override fun onFailure(call: Call, e: IOException) {
                callback(Result.failure(SDKException("Network error: ${e.message}")))
            }
        })
    }

    /**
     * Get list of products
     *
     * Example:
     * ```
     * client.getProducts(skip = 0, limit = 10) { result ->
     *     result.onSuccess { response ->
     *         println("Total products: ${response.total}")
     *         response.products.forEach { println(it.name) }
     *     }
     * }
     * ```
     */
    fun getProducts(
        skip: Int = 0,
        limit: Int = 10,
        category: String? = null,
        callback: (Result<ProductListResponse>) -> Unit
    ) {
        log("Fetching products...")

        val urlBuilder = HttpUrl.Builder()
            .scheme("http")
            .host(baseURL.removePrefix("http://").removePrefix("https://"))
            .addPathSegments("api/v1/products")
            .addQueryParameter("skip", skip.toString())
            .addQueryParameter("limit", limit.toString())

        category?.let { urlBuilder.addQueryParameter("category", it) }

        val request = Request.Builder()
            .url(urlBuilder.build())
            .addHeader("X-API-Key", apiKey)
            .get()
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onResponse(call: Call, response: Response) {
                response.use {
                    if (!response.isSuccessful) {
                        callback(Result.failure(SDKException("API error: ${response.code}")))
                        return
                    }

                    try {
                        val body = response.body?.string()
                        val result = gson.fromJson(body, ProductListResponse::class.java)
                        log("Products fetched", result)
                        callback(Result.success(result))
                    } catch (e: Exception) {
                        callback(Result.failure(SDKException("Failed to parse response: ${e.message}")))
                    }
                }
            }

            override fun onFailure(call: Call, e: IOException) {
                callback(Result.failure(SDKException("Network error: ${e.message}")))
            }
        })
    }

    private fun parseErrorMessage(body: String?): String {
        return try {
            val json = gson.fromJson(body, Map::class.java)
            json["detail"]?.toString() ?: "Unknown error"
        } catch (e: Exception) {
            body ?: "Unknown error"
        }
    }

    private fun log(message: String, data: Any? = null) {
        if (debug) {
            println("[BodyMeasurementSDK] $message ${data ?: ""}")
        }
    }
}

File: mobile-sdk/packages/ios/Sources/BodyMeasurementSDK/BodyMeasurementClient.swift
import Foundation
import UIKit

/// Main client for Body Measurement API
public class BodyMeasurementClient {
    private let apiKey: String
    private let baseURL: String
    private let session: URLSession
    private let debug: Bool

    public init(apiKey: String, baseURL: String = "http://localhost:8000", debug: Bool = false) {
        self.apiKey = apiKey
        self.baseURL = baseURL
        self.debug = debug
        self.session = URLSession.shared
    }

    /// Process single-person measurement
    public func processMeasurement(
        image: UIImage,
        productId: String? = nil,
        fitPreference: FitPreference = .regular,
        progress: ((Double) -> Void)? = nil,
        completion: @escaping (Result<MeasurementResponse, SDKError>) -> Void
    ) {
        guard let url = URL(string: "\(baseURL)/api/v1/measurements/process") else {
            completion(.failure(.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")

        let boundary = UUID().uuidString
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        request.httpBody = createMultipartBody(
            image: image,
            boundary: boundary,
            productId: productId,
            fitPreference: fitPreference
        )

        log("Sending measurement request...")

        session.dataTask(with: request) { [weak self] data, response, error in
            if let error = error {
                completion(.failure(.networkError(error)))
                return
            }

            guard let data = data else {
                completion(.failure(.noData))
                return
            }

            guard let httpResponse = response as? HTTPURLResponse else {
                completion(.failure(.invalidResponse))
                return
            }

            guard httpResponse.statusCode == 200 else {
                let errorMsg = self?.parseErrorMessage(from: data) ?? "Unknown error"
                completion(.failure(.apiError(httpResponse.statusCode, errorMsg)))
                return
            }

            do {
                let result = try JSONDecoder().decode(MeasurementResponse.self, from: data)
                self?.log("Measurement successful", result)
                completion(.success(result))
            } catch {
                completion(.failure(.decodingError(error)))
            }
        }.resume()
    }

    /// Get list of products
    public func getProducts(
        skip: Int = 0,
        limit: Int = 10,
        category: String? = nil,
        completion: @escaping (Result<ProductListResponse, SDKError>) -> Void
    ) {
        var components = URLComponents(string: "\(baseURL)/api/v1/products")!
        var queryItems = [
            URLQueryItem(name: "skip", value: "\(skip)"),
            URLQueryItem(name: "limit", value: "\(limit)")
        ]
        if let category = category {
            queryItems.append(URLQueryItem(name: "category", value: category))
        }
        components.queryItems = queryItems

        guard let url = components.url else {
            completion(.failure(.invalidURL))
            return
        }

        var request = URLRequest(url: url)
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")

        session.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(.networkError(error)))
                return
            }

            guard let data = data else {
                completion(.failure(.noData))
                return
            }

            do {
                let result = try JSONDecoder().decode(ProductListResponse.self, from: data)
                completion(.success(result))
            } catch {
                completion(.failure(.decodingError(error)))
            }
        }.resume()
    }

    // MARK: - Helper Methods

    private func createMultipartBody(
        image: UIImage,
        boundary: String,
        productId: String?,
        fitPreference: FitPreference
    ) -> Data {
        var body = Data()

        // Add image
        if let imageData = image.jpegData(compressionQuality: 0.8) {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"file\"; filename=\"photo.jpg\"\r\n".data(using: .utf8)!)
            body.append("Content-Type: image/jpeg\r\n\r\n".data(using: .utf8)!)
            body.append(imageData)
            body.append("\r\n".data(using: .utf8)!)
        }

        // Add api_key
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"api_key\"\r\n\r\n".data(using: .utf8)!)
        body.append("\(apiKey)\r\n".data(using: .utf8)!)

        // Add fit_preference
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"fit_preference\"\r\n\r\n".data(using: .utf8)!)
        body.append("\(fitPreference.rawValue)\r\n".data(using: .utf8)!)

        // Add product_id if provided
        if let productId = productId {
            body.append("--\(boundary)\r\n".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"product_id\"\r\n\r\n".data(using: .utf8)!)
            body.append("\(productId)\r\n".data(using: .utf8)!)
        }

        body.append("--\(boundary)--\r\n".data(using: .utf8)!)
        return body
    }

    private func parseErrorMessage(from data: Data) -> String {
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let detail = json["detail"] as? String {
            return detail
        }
        return String(data: data, encoding: .utf8) ?? "Unknown error"
    }

    private func log(_ message: String, _ data: Any? = nil) {
        if debug {
            print("[BodyMeasurementSDK] \(message)", data ?? "")
        }
    }
}
